nbconvert/preprocessors/execute.py

```diff
@@ -120,13 +125,13 @@ async def async_execute_cell(
         """
         # Copied and intercepted to allow for custom preprocess_cell contracts to be fullfilled
         self.store_history = store_history
-        cell, resources = self.preprocess_cell(cell, self.resources, cell_index)
+        cell, resources = await self.preprocess_cell(cell, self.resources, cell_index)
         # Apply rules from nbclient for where to apply execution counts
         if execution_count and cell.cell_type == 'code' and cell.source.strip():
             cell['execution_count'] = execution_count
         return cell, resources
 
-    def preprocess_cell(self, cell, resources, index, **kwargs):
+    async def preprocess_cell(self, cell, resources, index, **kwargs):
```

Comment:
other preprocessors'  `preprocess_cell` are not async. Should we make an async version of that method and call it from `async_execute_cell`?

Reply:
No, I think we just need to ensure it is async, I'm pushing another commit.

Reply:
it will most likely have consequences, because anyone overriding `preprocess_cell` in a non-async way will have surprises.

Reply:
I don't think so, since we now call `ensure_async(preprocess_cell())`, it can be async or not.

Reply:
as discussed in person, for people who override `preprocess_cell` and then call `super.preprocess_cell` in their implementation without `await` this should be an issue, but I need to check.

Reply:
It is an issue and it will break downstream projects like `nbdev` if `preprocess_cell` is made async in nbconvert

Reply:
We should try because I don't think it is a problem with the `ensure_async`.

Reply:
I believe it will be an issue to mix:

```
In [8]: class A:
   ...:     async def foo(self):
   ...:         print("bar")
   ...: 

In [9]: class B(A):
   ...:     def foo(self):
   ...:         print("foo")
   ...:         A.foo(self)
   ...:         print("!")
   ...: 

In [10]: B().foo()
foo
<ipython-input-9-33f3ae08d87f>:4: RuntimeWarning: coroutine 'A.foo' was never awaited
  A.foo(self)
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
!
```

Making the subclass aware of the parent being async but not being async itself on the override does weird things too:

```
In [12]: class B(A):
    ...:     def foo(self):
    ...:         print("foo")
    ...:         await A.foo(self)
    ...:         print("!")
    ...: 

In [13]: B().foo()
Out[13]: <coroutine object B.foo at 0x7f1d87179c40>
```